### PR TITLE
Allowing adding new token as unk token for gpt2 tokenizer

### PR DESCRIPTION
**Pull request by lsy641**
What does this PR do?

Fixes # (issue) #https://github.com/huggingface/transformers/issues/22414

## Before submitting
- [yes ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [yes] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [yes] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ no] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@ArthurZucker

### CODE REVIEW

**Review by lsy641 — COMMENTED** [review]
The third commit is just a commit after running "make style"

**Review by ArthurZucker — COMMENTED** [review]
Thanks for working on this.
What this does:
- Adding a new `unk_token`. This was not supported and can be classified as a bug. It gave a `None` index which breaks further encoding
- Tries to use the `added_tokens` when the `unk_token` is not part of the  `encoder`

I am okay with these changes!
IMO this fixes a behaviour and is backward compatible

**Comment by Narsil** [comment]
I think we should add some tests to clarify what behavior is modified an how.

It could be for just those 4 tokenizers, but still I think the effect of this PR is not entirely clear by just reading it.

**Comment by lsy641** [comment]
What do you need me to do to help this merge?

**Comment by ArthurZucker** [comment]
Hey! As mentioned in my last comments, the CI tests need to be all green 😉 Mostly `make fixup` should help you

File: src/transformers/tokenization_utils.py [hunk 1]
```diff
@@ -424,6 +424,10 @@ def _add_tokens(self, new_tokens: Union[List[str], List[AddedToken]], special_to
                 tokens_to_add.append(token)
                 if self.verbose:
                     logger.info(f"Adding {token} to the vocabulary")
+            elif token == self.unk_token and self.convert_tokens_to_ids(self.unk_token) is None and token not in tokens_to_add:
+                tokens_to_add.append(token)
+                if self.verbose:
+                    logger.info(f"Adding {token} to the vocabulary")
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils.py:430]: I tested locally and this is breaking a lot of models

**Pull request closed without merging**